doc/rst/meta/compiler-internals-doxygen/queries-impl.rst

```diff
@@ -0,0 +1,221 @@
+for some code that demonstrates avoiding recomputation.
+
+The depth-first traversal, combined with the need to recompute queries
+to check if their output has changed, means that dependencies are
+re-computed before the function that called them is re-computed. In
+other words, rather being recomputed *top-down* (from the top-level
+query down to its dependencies), dependencies are recomputed
+*bottom-up*. This has several implications for the way that context
+state should be managed (see the section on `state
+implications <#state-implications>`__).
+
+Once any of a query’s dependencies’ outputs is determined to have
+changed, there’s no use checking its other dependencies – any dependency
+changing is sufficient to need to recompute the whole dependent query.
+Thus, bottom-up recomputation suspends, and the dependent query is
+called. At this point, the query’s function runs as normal, its
+dependency queries are called as they are encountered in the C++ source
+code, which may invoke their own dependencies being computed, and so on.
+Thus, starting from the query-to-be-recomputed, we are once again back
+to a top-down mode of execution. However, at any point during this
+top-down execution, if the query needs to determine if a dependency
+needs to be recomputed, another bottom-up traversal will be needed.
+Thus, the execution of a query is a layering of top-down and bottom-up
+(re)computation.
+
+The recomputation-to-check-output requirement has one more practical
+effect. If a dependency is determined to have changed, this has occurred
+by re-running it. That means that when the parent query is called to in
+turn determine its new value, once it reaches the call to the changed
+dependency, the result will already be known, and the saved value will
+be used.
+
+The Preamble is Always Called
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One important thing to note is that, once you’ve decided that
+recomputation is necessary, “recomputing” the query is pretty much the
+same as computing it in the first place. At the time of writing,
+“recompute” is just a shim for calling the query function with the
+previously-supplied arguments. However, the query function’s body
+includes the preamble, and thus, the preamble is re-run whenever a
+query’s result is recomputed.
+
+The preamble does a few things, such as check if a query needs to be
+recomputed [2]_, push it onto a query stack (which records
+dependencies). If you are implementing logic that is sensitive to the
+kind of things that occur in the preamble, keep this in mind.
+
+Example of Avoiding Recomputation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here’s an example of when an input query changes, but a top-level query
+need not change. Suppose you’re reading a string from some input query,
+computing its length, and then performing an expensive computation with
+that somehow solely depends on the length.
+
+.. code:: c++
+
+   SomeDataStructure& expensiveQuery(Context* context) {
```

Comment:
did you mean to include a strLen argument here?

Also the QUERY_BEGIN calls need to include the function name.

Also something seems wrong with this example since I don't see why lengthQuery is run. Maybe you meant expensiveQuery to run lengthQuery on a string provided to it? Or something?

Reply:
I meant `lengthQuery` to invoke `inputQuery`, and `expensiveQuery` to invoke `lengthQuery`.